---
title: "Field Overlap"
output:
  html_document:
    df_print: paged
  pdf_document: default
---

```{r, results="hide", message=FALSE, warning=FALSE, fig.show="hide", echo=FALSE}
library(ggplot2)
library(plotly)
library(RSQLite)
library(reticulate)
library(gplots)
library(RColorBrewer)
```

# Field overlap

Using a python script to calculate minimum and maximum overlaps between all fields.
MinimumOverlap = commonGeneCount / max(geneCount)
MaximumOverlap = commonGeneCount / min(geneCount)

Keeping pairs with a maximum overlap greater than 0.5:


|Field A           |Field B       |Common genes|Min Overlap   |Max Overlap   |
|------------------|--------------|------------|--------------|--------------|
|PromoterFields:A  |CoexFields:G  |62          |0.04          |1.0           |
|PromoterFields:B  |CoexFields:G  |116         |0.08          |0.97          |
|PromoterFields:C  |CoexFields:F  |120         |0.12          |1.0           |
|PromoterFields:D  |CoexFields:F  |160         |0.16          |1.0           |
|PromoterFields:F  |CoexFields:G  |192         |0.13          |0.89          |
|PromoterFields:G  |CoexFields:F  |218         |0.22          |0.96          |
|MotifFields:B     |CoexFields:G  |82          |0.05          |0.85          |

## Genome hemispheres

We observe that only CoexFields have overlaps with other kinds of fields, and fields F and G in particular, which are the larger ones.
Taking a closer look at the rendering, we see that coexpression fields occupy different hemispheres of the nucleus.
The hemispheres are separated by a sea of "insulator".
This is nicely demonstrated by clusters A and B. They are different sectors of the Hat region, and they face away from each other and towards clusters F and G (respectively) of the bubble.
These two sectors of the hat are separated by the same insulator that separates the rest of the nucleus.

We also get a distinct "south pole" made by the centromeres.

Let's now combine these into a set of 4 clusters, using continent names as a consise indication of geometry:

CREATE TABLE Continents(Gene TEXT, Continent Text)

```
DROP TABLE IF EXISTS Continents;
CREATE TABLE Continents(Gene TEXT, Continent TEXT);
INSERT INTO Continents(Gene, Continent) SELECT Gene, 'Laurasia' FROM CoexFields WHERE Field = 'B' OR Field = 'C' OR Field = 'D' OR Field = 'G';
INSERT INTO Continents(Gene, Continent) SELECT Gene, 'Godwana' FROM CoexFields WHERE Field = 'A' OR Field = 'F';
INSERT INTO Continents(Gene, Continent) SELECT Gene, 'Antarctica' FROM CoexFields WHERE Field = 'E';
INSERT INTO Continents(Gene, Continent) SELECT Gene, 'Tethys' FROM Loci WHERE Gene NOT IN (SELECT Gene FROM Continents);
```
Here is how it looks:

```{r, message=FALSE, warning=FALSE}
sql <- "SELECT l.*, Continent FROM Loci l JOIN Continents c ON l.Gene = c.Gene"
fieldColors <- c("#FF0000", "#00FF00", "#0000FF", "#FFFF00", "#00FFFF", "#4488AA", "#FF00FF", "#000000")
conn <- dbConnect(RSQLite::SQLite(), "../../Results/yeast.sqlite")
continents <- dbGetQuery(conn, sql)
plot_ly(x=continents$x, y=continents$y, z=continents$z, type="scatter3d", mode="markers", text=continents$Gene, color=continents$Continent, colors=fieldColors)
dbDisconnect(conn)
```

And here are the respective overlaps. We have more fields overlapping now:

|Field           |Continent  |Common genes|Min Overlap|Max Overlap|
|----------------|-----------|------------|-----------|-----------|
|PromoterFields:A|Laurasia   |62          |0.027      |1.0        |
|PromoterFields:B|Laurasia   |116         |0.051      |0.97       |
|PromoterFields:F|Laurasia   |192         |0.08       |0.89       |
|MotifFields:B   |Laurasia   |82          |0.04       |0.85       |
|----------------|-----------|------------|-----------|-----------|
|PromoterFields:G|Godwana    |218         |0.21       |0.96       |
|PromoterFields:C|Godwana    |120         |0.12       |1.0        |
|PromoterFields:D|Godwana    |160         |0.16       |1.0        |
|----------------|-----------|------------|-----------|-----------|
|PromoterFields:E|Tethys     |98          |0.035      |0.57       |
|MotifFields:A   |Tethys     |55          |0.02       |0.96       |
|MotifFields:C   |Tethys     |76          |0.03       |0.78       |

## Continent statistics

### Box plots

10000 samples of pairwise coexpression score for each continent:

```{r}
continentCoexpression <- read.table(file = '../../Results/ContinentCoexpressionScoreSamples.tsv', sep = '\t', header = TRUE)
boxplot(continentCoexpression, outline=FALSE, las=2)
```

The important distinction is that of "sea versus land". Any combination of the 3 land continents will yield the same (better than average) coexpression score:

```{r}
continentCoexpression <- read.table(file = '../../Results/CombinedContinentCoexpressionScoreSamples.tsv', sep = '\t', header = TRUE)
boxplot(continentCoexpression, outline=FALSE, las=2)
```

### Statistical significance

| Compartment  | p-value   |
|-------------|------------|
| Godwana     | 1.18e-16   |
| Antarctica  | 1.03e-25   |
| Laurasia    | 7.02e-09   |
| Tethys      | 1.15e-54   |
| Pangaea     | 7.85e-14   |

### Detailed t-test results


#### Godwana

```{r}
t.test(continentCoexpression$Godwana, continentCoexpression$Random)
```

#### Antarctica

```{r}
t.test(continentCoexpression$Antarctica, continentCoexpression$Random)
```

#### Laurasia

```{r}
t.test(continentCoexpression$Laurasia, continentCoexpression$Random)
```

#### Tethys

```{r}
t.test(continentCoexpression$Tethys, continentCoexpression$Random)
```

#### Pangaea
```{r}
t.test(continentCoexpression$Pangaea, continentCoexpression$Random)
```

### Land continents are also significantly different with each other

```{r}
t.test(continentCoexpression$Godwana, continentCoexpression$Pangaea)
t.test(continentCoexpression$Antarctica, continentCoexpression$Pangaea)
t.test(continentCoexpression$Laurasia, continentCoexpression$Pangaea)
```

### Laurasia more dependent on the others

Note that Pangaea is actually more tightly coexpressed than its subset "Laurasia". Thus, Laurasian genes have a greater coexpression score with genes in Godwana and Antarctica, while the latter two are tighter than Pangaea.

### Antarctica significantly different to Pangaea

## Gene ontology enrichment

"Tethys" (coexpression insulator) cluster has a 2% PAdj for "Metabolit pathways".

## TightCommunities are continent-agnostic

These "Internal" gene sets have parts in all continents:

```
SELECT Chromosome, Continent, COUNT(Continent) FROM TightCommunities JOIN Continents 
ON TightCommunities.Gene = Continents.Gene 
JOIN Loci ON Loci.Gene = TightCommunities.Gene
GROUP BY Chromosome,Continent
```

|Chromosome|Continent |GeneCount|
|----------|----------|---------|
|2         |Godwana   |34       |
|4         |Antarctica|22       |
|4         |Laurasia  |26       |
|4         |Tethys    |13       |
|7         |Laurasia  |21       |
|7         |Tethys    |44       |
|10        |Antarctica|7        |
|10        |Laurasia  |27       |
|10        |Tethys    |21       |
|11        |Antarctica|22       |
|11        |Laurasia  |25       |
|11        |Tethys    |10       |
|13        |Godwana   |24       |
|13        |Tethys    |13       |
|14        |Laurasia  |68       |
|16        |Laurasia  |18       |
|16        |Tethys    |12       |

